lib/Infer/DataflowPruning.cpp

```diff
@@ -0,0 +1,388 @@
+      auto R1 = FindConstantRange(I->Ops[1], C);
+      return R0.multiply(R1);
+    }
+    case souper::Inst::And: {
+      auto R0 = FindConstantRange(I->Ops[0], C);
+      auto R1 = FindConstantRange(I->Ops[1], C);
+      return R0.binaryAnd(R1);
+    }
+    case souper::Inst::Or: {
+      auto R0 = FindConstantRange(I->Ops[0], C);
+      auto R1 = FindConstantRange(I->Ops[1], C);
+      return R0.binaryOr(R1);
+    }
+    case souper::Inst::Shl: {
+      auto R0 = FindConstantRange(I->Ops[0], C);
+      auto R1 = FindConstantRange(I->Ops[1], C);
+      return R0.shl(R1);
+    }
+    case souper::Inst::AShr: {
+      auto R0 = FindConstantRange(I->Ops[0], C);
+      auto R1 = FindConstantRange(I->Ops[1], C);
+      return R0.ashr(R1);
+    }
+    case souper::Inst::LShr: {
+      auto R0 = FindConstantRange(I->Ops[0], C);
+      auto R1 = FindConstantRange(I->Ops[1], C);
+      return R0.lshr(R1);
+    }
+    case souper::Inst::UDiv: {
+      auto R0 = FindConstantRange(I->Ops[0], C);
+      auto R1 = FindConstantRange(I->Ops[1], C);
+      return R0.udiv(R1);
+    }
+//     case souper::Inst::SDiv: {
+//       auto R0 = FindConstantRange(I->Ops[0], C);
+//       auto R1 = FindConstantRange(I->Ops[1], C);
+//       return R0.sdiv(R1); // unimplemented
+//     }
+    // TODO: Xor pattern for not, truncs and extends, etc
+    default: return result;
+  }
+}
+
+bool dataflow::ValueAnalysis::isInfeasible(souper::Inst* RHS) {
+  for (int i = 0; i < Inputs.size(); ++i) {
+    auto C = LHSValues[i];
+    if (C.hasValue()) {
+      auto CR = FindConstantRange(RHS, Inputs[i]);
+      if (!CR.contains(C.getValue())) {
+        return true;
+      }
+      auto KB = FindKnownBits(RHS, Inputs[i]);
+      if ((KB.Zero & C.getValue()) != 0 || (KB.One & ~C.getValue()) != 0) {
+        return true;
+      }
+
+      auto RHSV = Evaluate(RHS, Inputs[i]);
+      if (RHSV.hasValue()) {
+        if (C.getValue() != RHSV.getValue()) {
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+dataflow::DataflowPruningManager::DataflowPruningManager(souper::Inst* LHS, std::vector<Inst *> &Inputs, unsigned StatsLevel)
+: VA(LHS, generateInputSets(Inputs)), NumPruned(0), TotalGuesses(0) {
+
+  if (StatsLevel > 1) {
+    DataflowPrune= [this](Inst *I, std::vector<Inst *> &RI) {
+      TotalGuesses++;
+      if (VA.isInfeasible(I)) {
+        NumPruned++;
+        llvm::outs() << "Dataflow Pruned "
+          << NumPruned << "/" << TotalGuesses << "\n";
+        ReplacementContext RC;
+        RC.printInst(I, llvm::outs(), true);
+        return false;
+      }
+      return true;
+    };
+  } else if (StatsLevel == 1) {
+      DataflowPrune= [this](Inst *I, std::vector<Inst *> &RI) {
+      TotalGuesses++;
+      if (VA.isInfeasible(I)) {
+        NumPruned++;
+        return false;
+      }
+      return true;
+    };
+  } else {
+    DataflowPrune= [this](Inst *I, std::vector<Inst *> &RI) {
+      if (VA.isInfeasible(I)) {
```

Comment:
instead of the code you wrote here, you should write "return (!VA.isInfeasible(I))